jobs/buildhub/s3_inventory_to_kinto.py

```diff
@@ -129,25 +139,77 @@ def ends_with_date(prefix):
     manifest_content = json.loads(body.decode('utf-8'))
-    # Return keys of csv.gz files
     for f in manifest_content['files']:
-        yield f['key']
-
-
-async def download_csv(loop, s3_client, keys_stream, chunk_size=CHUNK_SIZE):
+        # Here, each 'f' is a dictionary that looks something like this:
+        #
+        #  {
+        #     "key" : "inventories/net-mozaw...f-b1a0-5fb25bb83752.csv.gz",
+        #     "size" : 7945521,
+        #     "MD5checksum" : "7454b0d773000f790f15b867ee152049"
+        #  }
+        #
+        # We yield the whole thing. The key is used to download from S3.
+        # The MD5checksum is used to know how to store the file on
+        # disk for caching.
+        yield f
+
+
+async def download_csv(
+    loop,
+    s3_client,
+    files_stream,
+    chunk_size=CHUNK_SIZE,
+    download_directory=CSV_DOWNLOAD_DIRECTORY,
+):
     """
     Download the S3 object of each key and return deflated data chunks (CSV).
     :param loop: asyncio event loop.
     :param s3_client: Initialized S3 client.
     :param keys_stream async generator: List of object keys for
     the csv.gz manifests.
     """
-    async for key in keys_stream:
-        key = 'public/' + key
-        logger.info('Fetching inventory piece {}'.format(key))
-        file_csv_gz = await s3_client.get_object(Bucket=BUCKET, Key=key)
+
+    # Make sure the directory exists if it wasn't already created.
+    if not os.path.isdir(download_directory):
+        os.mkdir(download_directory)
+
+    # Look for old download junk in the download directory.
+    too_old = 60 * 60 * 24 * 2  # two days
+    for file_path in glob.glob(os.path.join(download_directory, '*.csv.gz')):
+        age = time.time() - os.stat(file_path).st_mtime
+        if age > too_old:
+            logger.debug(
+                f'Delete old download file {file_path} '
+                f'({age} seconds old)'
+            )
+            os.remove(file_path)
+
+    async for files in files_stream:
+        # If it doesn't exist on disk, download to disk.
+        file_path = os.path.join(
+            download_directory,
+            files['MD5checksum'] + '.csv.gz'
+        )
+        # The file neither exists or has data.
+        if os.path.isfile(file_path) and os.stat(file_path).st_size:
+            logger.debug(f'{file_path} was already downloaded locally')
+        else:
+            key = 'public/' + files['key']
+            logger.info('Fetching inventory piece {}'.format(key))
+            file_csv_gz = await s3_client.get_object(Bucket=BUCKET, Key=key)
+            async with aiofiles.open(file_path, 'wb') as destination:
+                async with file_csv_gz['Body'] as source:
+                    while 'there are chunks to read':
+                        gzip_chunk = await source.read(chunk_size)
+                        if not gzip_chunk:
+                            break  # End of response.
+                        await destination.write(gzip_chunk)
```

Comment:
Here maybe you should try/except and delete the partially downloaded file when an error happens, or maybe you don't want to check the md5sum when resuming